File: app/lib/features/loans/data/guest_contact_model.dart
class GuestContactModel {
  GuestContactModel({
    required this.id,
    required this.name,
    this.phone,
    this.email,
    this.avatarColor = '#6C5CE7',
  });

  factory GuestContactModel.fromJson(Map<String, dynamic> j) => GuestContactModel(
        id: (j['_id'] ?? j['id']).toString(),
        name: j['name']?.toString() ?? '',
        phone: j['phone']?.toString(),
        email: j['email']?.toString(),
        avatarColor: j['avatarColor']?.toString() ?? '#6C5CE7',
      );

  final String id;
  final String name;
  final String? phone;
  final String? email;
  final String avatarColor;

  Map<String, dynamic> toJson() => {
        '_id': id,
        'name': name,
        if (phone != null) 'phone': phone,
        if (email != null) 'email': email,
        'avatarColor': avatarColor,
        'isGuest': true,
      };
}

File: backend/src/modules/loans/guest_contact.model.js
import mongoose from 'mongoose';

const guestContactSchema = new mongoose.Schema(
  {
    owner: { type: mongoose.Schema.Types.ObjectId, ref: 'User', required: true },
    name: { type: String, required: true },
    phone: { type: String, default: null },
    email: { type: String, default: null },
    avatarColor: { type: String, default: '#6C5CE7' },
    // The client's local UUID for this contact (used for idempotency and for
    // linking back guest loans to their contact after logout/re-login).
    clientId: { type: String, default: null },
    deletedAt: { type: Date, default: null },
  },
  { timestamps: true }
);

guestContactSchema.index({ owner: 1, updatedAt: 1 });
guestContactSchema.index({ clientId: 1 }, { unique: true, sparse: true });

export const GuestContact = mongoose.model('GuestContact', guestContactSchema);
